timelock-corruption/src/main/java/com/palantir/history/PaxosLogHistoryProvider.java

```diff
@@ -0,0 +1,89 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.history;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.sql.DataSource;
+
+import org.jdbi.v3.core.Jdbi;
+import org.jdbi.v3.sqlobject.SqlObjectPlugin;
+
+import com.palantir.history.models.CompletePaxosHistoryForNamespaceAndUsecase;
+import com.palantir.history.models.PaxosHistoryOnSingleNode;
+import com.palantir.history.sqlite.LocalHistoryLoader;
+import com.palantir.history.sqlite.LogVerificationProgressState;
+import com.palantir.paxos.NamespaceAndUseCase;
+import com.palantir.paxos.SqlitePaxosStateLogQueries;
+
+public class PaxosLogHistoryProvider {
+    private final DataSource dataSource;
+    private final LogVerificationProgressState logVerificationProgressState;
+    private final LocalHistoryLoader localHistoryLoader;
+    private final Jdbi jdbi;
+    private Map<NamespaceAndUseCase, Long> verificationProgressState = new ConcurrentHashMap<>();
+
+    private static final long INITIAL_PROGRESS = -1L;
+
+    public PaxosLogHistoryProvider(DataSource dataSource) {
+        this.jdbi = init(dataSource);
+        this.logVerificationProgressState = LogVerificationProgressState.create(dataSource);
+        this.localHistoryLoader = LocalHistoryLoader.create(dataSource);
+        this.dataSource = dataSource;
+    }
+
+    private Jdbi init(DataSource dataSource) {
+        Jdbi jdbi = Jdbi.create(dataSource).installPlugin(new SqlObjectPlugin());
+        jdbi.withExtension(SqlitePaxosStateLogQueries.class, SqlitePaxosStateLogQueries::createTable);
+        return jdbi;
+    }
+
+    private Set<NamespaceAndUseCase> getNamespaceAndUseCaseTuples() {
+        return jdbi.withExtension(SqlitePaxosStateLogQueries.class,
+                SqlitePaxosStateLogQueries::getAllNamespaceAndUseCaseTuples);
+    }
+
+    private Long getOrInsertVerificationState(NamespaceAndUseCase namespaceAndUseCase) {
+        Optional<Long> lastVerifiedSeq = logVerificationProgressState.getLastVerifiedSeq(
+                namespaceAndUseCase.namespace(), namespaceAndUseCase.useCase());
+        return lastVerifiedSeq.orElseGet(() -> insertVerificationState(namespaceAndUseCase));
+    }
+
+    private Long insertVerificationState(NamespaceAndUseCase namespaceAndUseCase) {
+        logVerificationProgressState.updateProgress(namespaceAndUseCase.namespace(), namespaceAndUseCase.useCase(),
+                INITIAL_PROGRESS);
+        return INITIAL_PROGRESS;
+    }
+
+    public List<CompletePaxosHistoryForNamespaceAndUsecase> getHistory() {
+        Map<NamespaceAndUseCase, Long> laseVerifiedSeqNamespaceAndUseCaseWise = getNamespaceAndUseCaseTuples().stream()
```

Comment:
nit: fix parameter name, `lastVerified...`

Probably enough to just use `lastVerifiedSequences`, the type name covers what the map is!